libraries/chain/steem_evaluator.cpp

```diff
@@ -1222,61 +1223,109 @@ void withdraw_vesting_evaluator::do_apply( const withdraw_vesting_operation& o )
-   FC_ASSERT( account.vesting_shares - account.delegated_vesting_shares >= o.vesting_shares, "Account does not have sufficient Steem Power for withdraw." );
+      FC_ASSERT( bal_obj != nullptr, "Account does not have sufficient Token Power for withdrawal." );
+      FC_ASSERT( bal_obj->vesting - bal_obj->delegated_vesting_shares >= o.vesting_shares, "Account does not have sufficient Token Power for withdrawal." );
 
-   FC_TODO( "Remove this entire block after HF 20" )
-   if( !_db.has_hardfork( STEEM_HARDFORK_0_20__1860 ) && !account.mined && _db.has_hardfork( STEEM_HARDFORK_0_1 ) )
-   {
-      const auto& props = _db.get_dynamic_global_properties();
-      const witness_schedule_object& wso = _db.get_witness_schedule_object();
+      if ( o.vesting_shares.amount == 0 )
+      {
+         FC_ASSERT( bal_obj->vesting_withdraw_rate.amount != 0, "This operation would not change the vesting withdraw rate." );
 
-      asset min_vests = wso.median_props.account_creation_fee * props.get_vesting_share_price();
-      min_vests.amount.value *= 10;
+         _db.modify( *bal_obj, [&]( account_regular_balance_object& a )
+         {
+            a.vesting_withdraw_rate = asset( 0, liquid_symbol.get_paired_symbol() );
+            a.next_vesting_withdrawal = time_point_sec::maximum();
+            a.to_withdraw = 0;
+            a.withdrawn = 0;
+         } );
+      }
+      else
+      {
+         _db.modify( *bal_obj, [&]( account_regular_balance_object& a )
+         {
+            int vesting_withdraw_intervals = STEEM_VESTING_WITHDRAW_INTERVALS;
 
-      FC_ASSERT( account.vesting_shares > min_vests || ( _db.has_hardfork( STEEM_HARDFORK_0_16__562 ) && o.vesting_shares.amount == 0 ),
-                 "Account registered by another account requires 10x account creation fee worth of Steem Power before it can be powered down." );
-   }
+            auto new_vesting_withdraw_rate = asset( o.vesting_shares.amount / vesting_withdraw_intervals, liquid_symbol.get_paired_symbol() );
 
-   if( o.vesting_shares.amount == 0 )
-   {
-      if( _db.has_hardfork( STEEM_HARDFORK_0_5__57 ) )
-         FC_ASSERT( account.vesting_withdraw_rate.amount  != 0, "This operation would not change the vesting withdraw rate." );
+            if( new_vesting_withdraw_rate.amount == 0 )
+               new_vesting_withdraw_rate.amount = 1;
 
-      _db.modify( account, [&]( account_object& a ) {
-         a.vesting_withdraw_rate = asset( 0, VESTS_SYMBOL );
-         a.next_vesting_withdrawal = time_point_sec::maximum();
-         a.to_withdraw = 0;
-         a.withdrawn = 0;
-      });
+            if( new_vesting_withdraw_rate.amount * vesting_withdraw_intervals < o.vesting_shares.amount )
+               new_vesting_withdraw_rate.amount += 1;
+
+            FC_ASSERT( a.vesting_withdraw_rate != new_vesting_withdraw_rate, "This operation would not change the vesting withdraw rate." );
```

Comment:
Do the math and the assertion outside of the modify, if possible. It minimizes performance impact if the assertion fails.